### PR TITLE
[improve] [txn] [PIP-160] Protocol changes and configuration changes for transaction batch log

### PR DESCRIPTION
**Pull request by poorbarcode**
Master Issue: #15370

### Motivation

see #15370

### Modifications

I will complete proposal #15370 with these pull requests( *current pull request is the step-2* ): 

1. Write the batch transaction log handler: `TxnLogBufferedWriter`
2. Configuration changes and protocol changes.
3. `MLPendingAckStore` and `MLTransactionLogImpl` support reading of batched logs.
4. `MLPendingAckStore` and `MLTransactionLogImpl` support the writing of batched logs and support dynamic configuration.
5. Append admin API for transaction batch log and docs( admin and configuration doc ).
6. Append metrics support for transaction batch log.

### Documentation

- [x] `doc-required` 
  
- [ ] `doc-not-needed` 
  
- [ ] `doc` 

- [ ] `doc-complete`

### CODE REVIEW

**Review by codelipenghui — COMMENTED** [review]
Please add an unit test to make sure the configuration is really work and make sure the default value is expected.

File: conf/broker.conf [hunk 1]
```diff
@@ -1535,4 +1535,28 @@ managedLedgerMaxUnackedRangesToPersistInZooKeeper=-1
...
+persistentUnackedRangesWithMultipleEntriesEnabled=false
+
+# Provide a mechanism allowing the Transaction Log Store to aggregate multiple records into a batched record and
+# persist into a single BK entry. This will make Pulsar transactions work more  efficiently, aka batched log.
+# see: https://github.com/apache/pulsar/issues/15370
+transactionLogBatchedWriteEnabled = false;
+# If enabled the feature that transaction log batch, this attribute means maximum log records count in a batch.
+transactionLogBatchedWriteMaxRecords= 512;
+# If enabled the feature that transaction log batch, this attribute means bytes size in a batch.
+transactionLogBatchedWriteMaxSize= 1024 * 1024 * 4;
```
Comment (codelipenghui) [inline conf/broker.conf:1547]: I think we can't user `1024 * 1024 * 4` here?

File: conf/broker.conf [hunk 2]
```diff
@@ -1535,4 +1535,28 @@ managedLedgerMaxUnackedRangesToPersistInZooKeeper=-1
...
 # If enabled, the maximum "acknowledgment holes" will not be limited and "acknowledgment holes" are stored in
 # multiple entries.
-persistentUnackedRangesWithMultipleEntriesEnabled=false
\ No newline at end of file
+persistentUnackedRangesWithMultipleEntriesEnabled=false
+
+# Provide a mechanism allowing the Transaction Log Store to aggregate multiple records into a batched record and
+# persist into a single BK entry. This will make Pulsar transactions work more  efficiently, aka batched log.
+# see: https://github.com/apache/pulsar/issues/15370
+transactionLogBatchedWriteEnabled = false;
```
Comment (codelipenghui) [inline conf/broker.conf:1543]:
```suggestion
transactionLogBatchedWriteEnabled=false;
```

Please check all.

**Comment by codelipenghui** [comment]
@poorbarcode Please update the PR title as the PR really changed.

**Comment by poorbarcode** [comment]
> @poorbarcode Please update the PR title as the PR really changed.

Already Fixed.

File: conf/broker.conf [hunk 3]
```diff
@@ -1535,4 +1535,28 @@ managedLedgerMaxUnackedRangesToPersistInZooKeeper=-1
...
+# see: https://github.com/apache/pulsar/issues/15370
+transactionPendingAckBatchedWriteEnabled = false;
+# If enabled the feature that transaction pending ack log batch, this attribute means maximum log records count in a
+# batch.
+transactionPendingAckBatchedWriteMaxRecords= 512;
+# If enabled the feature that transaction pending ack log batch, this attribute means bytes size in a batch.
+transactionPendingAckBatchedWriteMaxSize= 1024 * 1024 * 4;
+# If enabled the feature that transaction pending ack log batch, this attribute means maximum wait time(in millis) for
+# the first record in a batch.
+transactionPendingAckBatchedWriteMaxDelayInMillis= 1;
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply conf/broker.conf]: Already fixed.

File: conf/broker.conf (hunk 2 above)
Comment (poorbarcode) [reply conf/broker.conf]: Already fixed.

**Comment by poorbarcode** [comment]
>  Please add an unit test to make sure the configuration is really work and make sure the default value is expected.

This is a good way to solve the errors caused by carelessness. Thanks.

Already fixed.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 4]
```diff
@@ -2653,6 +2653,67 @@ public class ServiceConfiguration implements PulsarConfiguration {
...
+            category = CATEGORY_SERVER,
+            doc = "If enabled the feature that transaction log batch, this attribute means bytes size in a"
+                    + " batch, default 4m."
+    )
+    private int transactionLogBatchedWriteMaxSize = 1024 * 1024 * 4;
+
+    @FieldContext(
+            category = CATEGORY_SERVER,
+            doc = "If enabled the feature that transaction log batch, this attribute means maximum wait time(in millis)"
+                    + " for the first record in a batch, default 1."
```
Comment (liangyepianzhou) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2682]:
```suggestion
                    + " for the first record in a batch, default 1 millisecond."
```

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 5]
```diff
@@ -2653,6 +2653,67 @@ public class ServiceConfiguration implements PulsarConfiguration {
...
+            category = CATEGORY_SERVER,
+            doc = "If enabled the feature that transaction pending ack log batch, this attribute means bytes size in"
+                    + " a batch, default 4m."
+    )
+    private int transactionPendingAckBatchedWriteMaxSize = 1024 * 1024 * 4;
+
+    @FieldContext(
+            category = CATEGORY_SERVER,
+            doc = "If enabled the feature that transaction pending ack log batch, this attribute means maximum wait"
+                    + " time(in millis) for the first record in a batch, default 1."
```
Comment (liangyepianzhou) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java:2713]:
```suggestion
                    + " time(in millisecond) for the first record in a batch, default 1 millisecond."
```

**Review by liangyepianzhou — APPROVED** [review]
Good work! Leave a few comments.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java (hunk 4 above)
Comment (poorbarcode) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: Already fixed.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java (hunk 5 above)
Comment (poorbarcode) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: Already fixed.

**Review by congbobo184 — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]
LGTM,
Another question: Should we add this configuration in standalone.conf?

**Comment by poorbarcode** [comment]
> LGTM, Another question: Should we add this configuration in standalone.conf?

Already fixed. Thanks

**Comment by momo-jun** [comment]
Hi, @poorbarcode, will you add or update relevant docs in a follow-up PR?

**Pull request merged**